napari/components/viewer_model.py

```diff
@@ -97,155 +93,26 @@ def __init__(
             )
+        self.events.palette()  # TODO: can we just make this events.theme?
 
-    @property
-    def grid_size(self):
-        """tuple: Size of grid
-        """
-        return self._grid_size
-
-    @grid_size.setter
-    def grid_size(self, grid_size):
-        if np.all(self.grid_size == grid_size):
-            return
-        self._grid_size = grid_size
-        self.reset_view()
-        self.events.grid()
-
-    @property
-    def status(self):
-        """string: Status string
-        """
-        return self._status
-
-    @status.setter
-    def status(self, status):
-        if status == self.status:
-            return
-        self._status = status
-        self.events.status(text=self._status)
-
-    @property
-    def help(self):
-        """string: String that can be displayed to the
-        user in the status bar with helpful usage tips.
-        """
-        return self._help
-
-    @help.setter
-    def help(self, help):
-        if help == self.help:
-            return
-        self._help = help
-        self.events.help(text=self._help)
-
-    @property
-    def title(self):
-        """string: String that is displayed in window title.
-        """
-        return self._title
-
-    @title.setter
-    def title(self, title):
-        if title == self.title:
-            return
-        self._title = title
-        self.events.title(text=self._title)
-
-    @property
-    def interactive(self):
-        """bool: Determines if canvas pan/zoom interactivity is enabled or not.
-        """
-        return self._interactive
-
-    @interactive.setter
-    def interactive(self, interactive):
-        if interactive == self.interactive:
-            return
-        self._interactive = interactive
-        self.events.interactive()
-
-    @property
-    def cursor(self):
-        """string: String identifying cursor displayed over canvas.
-        """
-        return self._cursor
-
-    @cursor.setter
-    def cursor(self, cursor):
-        if cursor == self.cursor:
-            return
-        self._cursor = cursor
-        self.events.cursor()
-
-    @property
-    def cursor_size(self):
-        """int | None: Size of cursor if custom. None is yields default size
-        """
-        return self._cursor_size
-
-    @cursor_size.setter
-    def cursor_size(self, cursor_size):
-        if cursor_size == self.cursor_size:
-            return
-        self._cursor_size = cursor_size
-        self.events.cursor()
-
-    @property
-    def active_layer(self):
-        """int: index of active_layer
-        """
-        return self._active_layer
-
-    @active_layer.setter
-    def active_layer(self, active_layer):
-        if active_layer == self.active_layer:
-            return
-
-        if self._active_layer is not None:
-            self.keymap_providers.remove(self._active_layer)
+    def _on_active_layer_set(self, active_layer):
+        # FIXME: we need access to the OLD value
```

Comment:
Why isn't `self._active_layer` still the old value? i.e. if the arg `active_layer` is the new value and this method is getting called as a `super` first thing in the setter then when it gets called the old value is still available right and the new value hasn't been set yet?

Reply:
> this method is getting called as a super first thing in the setter

That's not what this method is doing. It's not meant as a setter, (that's what the `Property[..., ..., setter]` is for.  This method was there in case you needed to do complex things like emit different events and/or call additional internal methods after setting.  If you just want to modify the setting process, I think we should be using a `Property`